proposal/new-function-and-values-api.d.ts

```diff
@@ -0,0 +1,868 @@
+/**
+ * # New Function and Values API: Draft 1
+ *
+ * *([Issue](https://github.com/sass/sass/issues/2510))*
+ *
+ * ## Background
+ *
+ * > This section is non-normative.
+ *
+ * Sass's current JavaScript API was inherited from Node Sass, which developed
+ * over time in an _ad hoc_ manner. The functions API in particular has a number
+ * of notable issues:
+ *
+ * - Values are mutable. This goes against the grain of how Sass values work,
+ *   and makes it difficult to implement the API efficiently in Dart Sass.
+ *
+ * - It's hard to verify argument types with assertions.
+ *
+ * - It's hard to follow Sass's conventions for dealing with objects, such as:
+ *
+ *   - Treating `false` and `null` as falsey and everything else as truthy.
+ *   - Treating all values as lists, and treating maps as lists of pairs.
+ *   - Treating empty lists and empty maps as the same value.
+ *   - Using 1-based indexes for lists and strings, and negative indexes to
+ *     count from the back.
+ *   - Indexing the Unicode codepoints of strings rather than UTF-16 code units.
+ *   - There's no representation of first-class function values.
+ *
+ * - There's no representation of an argument list value that may contain
+ *   keyword arguments.
+ *
+ * - Maps are represented as a list of pairs, without any way of accessing a
+ *   value using its key other than iterating through the entire list.
+ *
+ * - Numbers represent their units as a single string in an undocumented format,
+ *   making it very difficult to work with them if they have more than just a
+ *   single numerator unit.
+ *
+ * - The `this` context includes a bunch of undocumented information without a
+ *   clear use-case.
+ *
+ * This proposal for a new Function and Values API solves these problems. It is
+ * heavily based on Dart Sass's [Dart `Value` API].
+ *
+ * [Dart `Value` API]: https://pub.dev/documentation/sass/latest/sass/Value-class.html
+ */
+
+/** API */
+
+import {OrderedMap} from 'immutable';
+
+import './new-js-api';
+
+type CustomFunctionCallback = (args: Value[]) => Value;
+
+/**
+ * This definition updates the one in the [New JavaScript API proposal].
+ *
+ * [New JavaScript API proposal]: ./new-js-api.d.ts
+ */
+declare module './new-js-api' {
+  interface Options<sync extends 'sync' | 'async'> {
+    /**
+     * When the compiler encounters a function with a signature that does not
+     * match that of a built-in function, but matches a key in this map, it must
+     * call the associated `CustomFunctionCallback`.
```

Comment:
```suggestion
     * call the associated `CustomFunctionCallback` and return its result.
```

Reply:
I think it's also worth explicitly saying "throw an error if this returns anything other than a `Value`". Normally we just say "failing to match type declarations produces undefined behavior", but this specific call feels likely enough to do something wrong (such as use an old-API function) that I think it's worth being explicit.